kythe/java/com/google/devtools/kythe/analyzers/java/AutoValuePlugin.java

```diff
@@ -0,0 +1,296 @@
+    // Check if class was generated by the AutoValueProcessor
+    if (!classDef.getModifiers().getAnnotations().stream()
+        .filter(
+            // Class decorated by @Generated(value = "...AutoValueProcessor")
+            ann ->
+                GENERATED_ANNOTATIONS.contains(annotationName(ann))
+                    && AUTO_VALUE_PROCESSOR.equals(annotationLiteralValue(ann, "value")))
+        .findFirst()
+        .isPresent()) {
+      return false;
+    }
+
+    // Find @AutoValue superclass definition.
+    JCClassDecl autoValueClass = null;
+    ClassType classType = (ClassType) classDef.type;
+    while (autoValueClass == null && classType.supertype_field instanceof ClassType) {
+      classType = (ClassType) classType.supertype_field;
+      JCTree superTypeTree = javacTrees.getTree(classType.asElement());
+      if (superTypeTree instanceof JCClassDecl) {
+        JCClassDecl cls = (JCClassDecl) superTypeTree;
+        if (getAnnotation(AUTO_VALUE_ANNOTATION, cls.getModifiers().getAnnotations()).isPresent()) {
+          autoValueClass = cls;
+        }
+      }
+    }
+    if (autoValueClass == null) {
+      return false;
+    }
+
+    // @AutoValue class generates the implementation class.
+    Optional<KytheNode> autoNode = kytheGraph.getNode(autoValueClass.sym);
+    Optional<KytheNode> genNode = kytheGraph.getNode(classDef.sym);
+    if (autoNode.isPresent() && genNode.isPresent()) {
+      entrySets.emitEdge(autoNode.get().getVName(), EdgeKind.GENERATES, genNode.get().getVName());
+    }
+
+    Map<String, JCMethodDecl> propMethods = new HashMap<>(); // property_name -> getter method
+    Map<String, VName> getterNodes = new HashMap<>(); // property_name -> Kythe getter VName
+    Map<String, VName> properties = new HashMap<>(); // property_name -> Kythe property node VName
+
+    // Find potential propMethods in @AutoValue class.
+    for (JCTree def : autoValueClass.getMembers()) {
+      if (def instanceof JCMethodDecl) {
+        JCMethodDecl method = (JCMethodDecl) def;
+        if (method.getModifiers().getFlags().contains(Modifier.ABSTRACT)
+            && method.getParameters().isEmpty()
+            && !isVoid(method.getReturnType())) {
+          Optional<KytheNode> kytheNode = kytheGraph.getNode(method.sym);
+          if (kytheNode.isPresent()) {
+            String name = method.sym.getSimpleName().toString();
+            // TODO(schroederc): possibly reuse getter method as "property" node or change node kind
+            EntrySet propNode =
+                entrySets.newNode("property").addSignatureSalt(kytheNode.get().getVName()).build();
+            propNode.emit(entrySets.getEmitter());
+            entrySets.emitEdge(
+                kytheNode.get().getVName(), EdgeKind.PROPERTY_READS, propNode.getVName());
+
+            properties.put(name, propNode.getVName());
+            propMethods.put(name, method);
+            getterNodes.put(name, kytheNode.get().getVName());
+          }
```

Comment:
Maybe define an @AutoValue for {propNode.getVName(), method, kytheNode.get().getVName()) and have a single Map instead of 3. Then you'd be able to factor this large method out a bit more.